### PR TITLE
Add support for LogAppend timestamp type

### PR DESCRIPTION
**Pull request by zimin2000**
PR as requested in https://github.com/liftoffio/liftoff/pull/4320 . It out local version for https://github.com/Shopify/sarama/pull/1258 and https://github.com/Shopify/sarama/pull/1259 .

### CODE REVIEW

File: message.go [hunk 1]
```diff
@@ -57,6 +60,9 @@ func (m *Message) encode(pe packetEncoder) error {
 	pe.putInt8(m.Version)
 
 	attributes := int8(m.Codec) & compressionCodecMask
+	if m.LogAppendTime {
+		attributes |= timestampTypeMask
```
Comment (zimin2000) [inline message.go:65]: by @mikejquinn:
> Does this make sense to be set by a producer? I thought this was a server config setting.

This change is done for the sarama tests to pass. The tests use `encode`/`decode` to simulate message delivery. So, yes, normally the flag is only set by the Kafka and should not be considered there.

I think I could change the tests to overcome this, but I am not sure it worth the efforts.

File: message.go [hunk 2]
```diff
@@ -24,6 +24,8 @@ const (
 	CompressionLZ4    CompressionCodec = 3
 )
 
+const timestampTypeMask = 0x08
```
Comment (zimin2000) [inline message.go]: by @mikejquinn:
> Move this below the CompressionCodec.String function and CompressionLevelDefault const.

File: consumer.go [hunk 3]
```diff
@@ -516,13 +520,17 @@ func (child *partitionConsumer) parseRecords(batch *RecordBatch) ([]*ConsumerMes
 		if offset < child.offset {
 			continue
 		}
+		timestamp := batch.FirstTimestamp.Add(rec.TimestampDelta)
```
Comment (cespare) [inline consumer.go]: Not that this computation is particularly expensive but I would probably declare `var timestamp time.Time` and then do one or the other in an if/else.

**Review by cespare — APPROVED** [review]
So this was just a completely missing feature on their end?

**Comment by zimin2000** [comment]
I was surprised myself that such an important feature can be missing, but it seems to be the case.

**Pull request merged**